surveys/promise-rejections/survey.md

```diff
@@ -0,0 +1,152 @@
+# Node.js Promise reject use case survey
+
+Today, Node.js handles unhandled rejections by emitting a deprecation warning to stderr. The warning shows the stack where the rejection happened, and states that in future Node.js versions unhandled rejections will result on Node.js exitting with non-zero status code. We intend to remove the deprecation warning, replacing it with a stable behavior which might be different from the one described on the deprecation warning. We're running this survey to better understand how Node.js users are using Promises and how they are dealing with unhandled rejections today, so we can make an informed decision on how to move forward.
+
+## What is a Promise rejection?
+
+A Promise rejection indicates that something went wrong while executing a Promise or an async function. Rejections can occur on several situations: thorwing inside an async function or a `Promise` executor/then/catch/finally callback, when calling the `reject` callback of an `executor`, or when calling `Promise.reject`.
+
+```js
+
+Promise.reject()  // This will result in a rejection
+
+new Promise((fulfill, reject) => {
+  reject();  // This will result in a rejection
+});
+
+new Promise(() => {
+  throw new Error();  // This will result in a rejection
+}).then(() => {
+  throw new Error();  // This will result in a rejection
+}, () => {
+  throw new Error();  // This will result in a rejection
+}).catch(() => {
+  throw new Error();  // This will result in a rejection
+}).finally(() => {
+  throw new Error();  // This will result in a rejection
+});
+
+async function () {
+  throw new Error();  // This will result in a rejection
+}
+
+async function () {
+  if (a === undefined)
+    a();  // This will result in a rejection
+}
+```
+
+## What is an unhandled rejection?
+
+Rejections are designed to be handled asynchronously. In JavaScript, Promises can be handled when awaiting for a promise inside a try/catch block, or via .catch calls:
+
+
+```js
+async function foo() {
+  throw new Error();
+}
+
+foo().catch(() => console.error("an error occured"));  // 1. Handled
+
+try {
+  await foo();  // 2. Handled
+} catch(e) {
+  console.error("an error occured")
+}
+
+foo(); // 3. Unhandled, but execution continues
+
+const rejected = foo(); // 4. Unhandled on current event loop tick
```

Comment:
I would prefer "task" or something other than "event loop tick" which is a bit vague in terms of terminology.

Reply:
Hmm, I find “event loop tick” precise and “task” to me means “promise-returning function” :-)

Reply:
I'm honestly not sure what "event loop tick" means, I've never thought of it as a loop and I doubt users care about [internals](https://nodejs.org/fa/docs/guides/event-loop-timers-and-nexttick/) (even though that article isn't exactly accurate it's the best one I know).

If you think it's clearer I will concede, it's possible the vague terminology is a lot clearer than task/microtask.

Reply:
stuff i send to newcomers in irc include https://nodesource.com/blog/understanding-the-nodejs-event-loop , https://www.youtube.com/watch?v=8aGhZQkoFbQ , https://www.youtube.com/watch?v=cCOL7MC4Pl0 - "event loop" is generally a well known concept from the browser. i've heard both "tick" and "turn" to describe an iteration of the event loop.

Reply:
I rarely (if every) hear Node.js developers refer to it as `task`. It's worth noting that this survey is targeted to Node.js developers, who may not be familiar with ECMAScript spec terminology.

I like `turn` better than `tick`, thanks @ljharb, I'll update this comment.